src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -1572,7 +1533,7 @@ void EmitWhenHasCapture()
                     additionalDeclarations.Add("int matchLength = 0;");
                     writer.WriteLine($"matchLength = base.MatchLength({capnum});");
 
-                    bool caseInsensitive = IsCaseInsensitive(node);
+                    bool caseInsensitive = (node.Options & RegexOptions.IgnoreCase) != 0;
```

Comment:
Don't we need to avoid using the source generator if there's an ignore-case backreference?  That means this ignore-case logic should be removed from backreferences in the source generator (it would remain in RegexCompiler), and we need to update the logic that determines whether the source generator emits an implementation to check for whether there's an IgnoreCase Backreference.

Given that tests aren't failing because of this, seems we're also missing test cases.

Reply:
I'm not sure I'm following here. I know in the past you and I chatted about not supporting ignoreCase backreferences in the source generator, but that was because we wanted to use the same Regex case equivalence table for backreferences, and that is not what I'm doing yet here. For backreferences with Ignore case (in Interpreted, Compiled, and Source generated engines), I'm currently just falling back to do what we used to do which is walking the previously matched back reference character by character and calling ToLower() on both sides using CurrentCulture for casing comparisons. I understand this is not optimal, but it is also not worse than what we had. My thinking was that we could start with this, and then look for ways to optimize backreference IgnoreCase comparisons.

For example, this is a small snippet that would be emitted by the source generated engine for a backreference with IgnoreCase:

```c#
                        // Match the same text as matched by the 1st capture group.
                        {
                            pos++;
                            slice = inputSpan.Slice(pos);
                            
                            // If the 1st capture group hasn't matched, the backreference doesn't match.
                            if (!base.IsMatched(1))
                            {
                                UncaptureUntil(0);
                                return false; // The input didn't match.
                            }
                            
                            // Get the captured text.  If it doesn't match at the current position, the backreference doesn't match.
                            matchLength = base.MatchLength(1);
                            if (slice.Length < matchLength)
                            {
                                UncaptureUntil(0);
                                return false; // The input didn't match.
                            }
                            
                            matchIndex = base.MatchIndex(1);
                            for (int i = 0; i < matchLength; i++)
                            {
                                if (textInfo.ToLower(inputSpan[matchIndex + i]) != textInfo.ToLower(slice[i]))
                                {
                                    UncaptureUntil(0);
                                    return false; // The input didn't match.
                                }
                            }
                            
                            pos += matchLength;
                            slice = inputSpan.Slice(pos);
                        }
```

We do have tests for backreferences in IgnoreCase scenarios, and they all pass in all engines after my changes. Does that more or less answer the question or am I missing something?

Reply:
> I'm currently just falling back to do what we used to do which is walking the previously matched back reference character by character and calling ToLower() on both sides using CurrentCulture for casing comparisons. I understand this is not optimal, but it is also not worse than what we had.

It means we now have two different definitions for how casing works.  Doesn't that lead to inconsistencies as to whether things are considered case-equivalent, based on whether we're matching against the pattern or against other input in a backreference?  I don't think that's viable.

Reply:
Note, too, that one of our goals was to avoid using the match-time culture, to avoid inconsistencies resulting from the culture changing between construction and match. This goes against that.

Reply:
Our Casing equivalence table already is matching CurrentCulture.TextInfo.ToLower() behavior, so the only edge case where we would have inconsistencies is if you are comparing two characters that both lowercase to the same character, but have no other relationship between each other (which AFAIK is the case only for '\u0130' and 'I'). For this edge case, we would not consider a match in normal circumstances but would consider a match in a backreference case, but given it is only one small edge case, I thought it was ok to either accept this small inconsistency or probably to instead just emit code that handles this small edge case. I basically saw more value of supporting Source generator for patterns with ignorecase and backreferences than not doing so just for that small edge case, but I'm open to suggestions.

Reply:
> Note, too, that one of our goals was to avoid using the match-time culture, to avoid inconsistencies resulting from the culture changing between construction and match. This goes against that.

While I agree that this means we are using the match-time culture, it is slightly different than our previous concern which was that you may have a culture mismatch between pattern parsing and match-time, but with what I'm doing here, there is no chance for an inconsistency since both the backreference and the text being tested use the same match-time culture. Note that this problem would still exist even if we were using the Regex casing equivalence table for backreferences since we need to evaluate the culture in order to know which special mapping behaviors to use (unless we decide to cache the culture that was detected at creation time, which wouldn't be ideal)

Reply:
> unless we decide to cache the culture that was detected at creation time, which wouldn't be ideal

That's what I thought we'd do in the interpreter and compiler (specifically which rules to use). What's not ideal about it?

Reply:
To be clear, I'm not 100% against having special rules specifically for backrefs, but it is a very subtle inconsistency that affects behavior, something we really want to avoid in general. If we do it, I think we need to ensure the culture used for construction is cached to use for match. I'm also not clear on what you'd propose to do if we make your other proposed change of also including ToUpper in the case equivalence function.